integration/cli_test.go

```diff
@@ -306,98 +395,184 @@ func Test3LOFlow(t *testing.T) {
+			"fetch-3lo-cached.golden",
+			false,
+		},
+		{
+			"fetch; 3lo loopback insert expired token into cache",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", CONSET_PAGE_TIMEOUT, "--consentPageInteractionTimeoutUnits", CONSET_PAGE_TIMEOUT_UNITS},
+			"fetch-3lo-loopback.golden",
+			false,
+		},
+		{
+			"fetch; 3lo loopback cached; token expired",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", CONSET_PAGE_TIMEOUT, "--consentPageInteractionTimeoutUnits", CONSET_PAGE_TIMEOUT_UNITS},
+			"fetch-3lo-loopback.golden",
+			false,
+		},
+		{
+			"fetch; 3lo loopback cached; refresh expired token",
+			[]string{"fetch", "--scope", "pubsub", "--credentials", "integration/fixtures/fake-client-secrets-expired-token-3lo-loopback.json", "--refresh",
+				"--disableAutoOpenConsentPage",
+				"--consentPageInteractionTimeout", CONSET_PAGE_TIMEOUT, "--consentPageInteractionTimeoutUnits", CONSET_PAGE_TIMEOUT_UNITS},
+			"fetch-3lo-cached.golden",
+			false,
+		},
 	}
 
-	process3LOOutput := func(output string) string {
-		re := regexp.MustCompile("redirect_uri=http%3A%2F%2Flocalhost%3A\\d+")
-		match := re.FindString(output)
-		output = strings.Replace(output, match, "redirect_uri=http%3A%2F%2Flocalhost", 1)
-		return removeCodeChallenge(output)
+	type LoopbackLogicState struct {
+		quitRetry bool
+		cred      *testFile
+	}
+	loopbackLogic := func() (func(tc *testCase, loopbackAddr string) error, func(loopbackAddr string)) {
+		var ll *LoopbackLogicState
+		const (
+			CODE_AND_STATE   = "/?state=state&code=4/gwEhAq4N7tdTj4ZStstQgaDAUpcoceoFSEPmSsoWEKVZoYSn6URLVEw"
+			QUOTED_LOCALHOST = "\"http://localhost\""
+			TEMP_CRED_FILE   = "temp-cred-file.json"
+		)
+
+		preLogic := func(tc *testCase, loopbackAddr string) error {
+			ll = &LoopbackLogicState{}
+
+			// searching for credentials filename.
+			f := getCredentialsFileName(tc)
+			if f == "" {
+				return fmt.Errorf("Credentials file is missing. Please add to test arguments.")
+			}
+			changeCredentialsFileName(tc, TEMP_CRED_FILE)
+
+			// Loading credentials file
+			origFile := newFixture(t, path.Base(f))
+			origFile.saveAs(TEMP_CRED_FILE)
+
+			// Loading credentials file temp copy
+			(*ll).cred = newFixture(t, TEMP_CRED_FILE)
+			fileContent := (*ll).cred.load()
+
+			// Modifiying credentials file temp copy: redirect uri
+			re := regexp.MustCompile(QUOTED_LOCALHOST)
+			match := re.FindString(fileContent)
+			newContent := strings.Replace(fileContent, match, "\""+loopbackAddr+"\"", 1)
+			(*ll).cred.write(newContent)
+
+			// Start loopback logic.
+			go func() {
+				timer := time.AfterFunc(CONSET_PAGE_DURATION, func() {
```

Comment:
Using timer here seems a bit overkill. I would convert the for loop condition to take an integer counter or something, so it ends after 30 loops max. Also, is this to safeguard against the scenario where postLogic doesn't get triggered? i.e. CMD crashes?

Reply:
Thanks for your suggestion and please excuse me if I am over explaining. 

Yes, it does seem a little overkilled, but I do prefer it over a max try counter. Imagine that we are running the tests in a very slow machine. It takes 40 seconds to start serving the POST request and so all tests are failing for this flow. To fix it all is needed is to change the CONSENT_[CONSTANTS] - all of them. If a counter is used, then the SWE will need to dive into more complicated code to change the max try.  

In theory, the safety-net should never get trigger. As you pointed out, it should only get triggered when the postLogic is not configured correctly or when a crash occurs. If you believe I am being too cautious we can remove the timeout in its entirety. So there will be no need for a max counter either. Just  a friendly reminder: the loop is running async.  

Thanks!

Reply:
Sg, we can keep it as is for now. I general I don't think we need to be as cautious with test code since we are running it in a controlled environment, so if it passes once, it should always pass, and not trigger rare edge cases - and the consequence of a test failing on rare edge cases is also low. For example, the current tests before loopback support was added already has some shortcuts such as using port 8080 for mock server, and we have not ran into port collision problem yet, but we should probably fix that at some point too.